### PR TITLE
optimize `ExpAtom`

### LINKED ISSUE #254: Massive RAM for moderate problem
I have an example of a moderately sized problem where problem formulation is consuming massive memory. I have no idea why this happens. Is there something inherently difficult to this kind of problem? Or is this a bug in Convex.jl?

Note that this is not an issue with any solver. The RAM spike occurs at the problem formulation step, the line `problem = maximize(...)` below.

To run this code, you need to load the variables `N, Adj` from the `.jld2` file (using JLD2.jl) here:

https://github.com/cossio/cvx_example_data

`N` is an float 3-dimensional Array, and `Adj` a sparse matrix.

**WARNING:** The following code took so much memory that I had to reboot my laptop. Run this in a contained environment where you can constrain memory consumption. (For example, using https://github.com/pshved/timeout, start julia with `timeout -m 4000000 julia` to limit memory usage to 4GB.)

```julia
x = Variable(2730)
S, V, T = size(N)
lN = log.(N)
M = vec(sum(N[:,:,2:T]; dims=(2,3)))
H = Adj * x
problem = maximize(-sum(logsumexp(lN[:,v,t] - H) for t = 1:T-1, v = 1:V) - dot(M, H), [x ≥ -1e2, x ≤ 1e2])
solve!(problem, ECOSSolver())
```

**UPDATE:** I'm now getting the RAM spike at the solve phase, instead of the formulation phase. Both ECOS and SCS have the same problem.

### PR DESCRIPTION
**Pull request by ericphanson**
closes #254 

Before: on the original problem, today I OOM'd on my 16 GB macbook pro. On [previous attempts](https://github.com/jump-dev/Convex.jl/issues/254#issuecomment-1636944289), I got
```julia
1303.736478 seconds (11.94 M allocations: 191.588 GiB, 92.49% gc time)
```

On this branch:
```julia
  5.474096 seconds (1.13 M allocations: 216.129 MiB, 20.82% gc time)
```

The issue is formulating 33378 IndexAtoms and so forth. In general in Convex we must never do scalar operations if we hope to formulate things performantly (unless we move everything to a scalar representation like MOI).

Here, I couldn't see how to avoid the scalar operations, so I lower to MOI to do them on that level, which is efficient.

I filed #614 to track this issue more generally.

### CODE REVIEW

File: src/atoms/ExpAtom.jl [hunk 1]
```diff
@@ -34,9 +34,37 @@ function new_conic_form!(context::Context{T}, e::ExpAtom) where {T}
...
+    # for i in 1:m, j in 1:n
+    #     f = vcat(x[i, j], 1, z[i, j])
+    #     add_constraint!(context, GenericConstraint{MOI.ExponentialCone}(f))
+    # end
+    # return conic_form!(context, z)
+    # This is slow, since we are indexing on the Convex side, and convex is based around
+    # vector/matrix operations. We don't want to produce n*m IndexAtoms!
+    # Instead, we will drop to the MOI level to implement this in terms of scalar operations.
+    # First, we will get `x` as an MOI.VectorAffineFunction
+    x_tape = conic_form!(context, x)
```
Comment (ericphanson) [inline src/atoms/ExpAtom.jl:47]: our tests seem to say it's fine, but I don't actually 100% remember if this will always be a `SparseTape{T}` like I expect here... I wonder if in some edge cases it can be something else, like a `Vector{T}`. I need to double-check.

File: src/problem_depot/problems/exp.jl [hunk 2]
```diff
@@ -22,6 +22,19 @@
         @test evaluate(exp(y)) ≈ 1 atol = atol rtol = rtol
     end
 
+    # Test for constant `exp` (#613)
+    y = Variable()
+    x = constant([1,2,3])
```
Comment (ericphanson) [inline src/problem_depot/problems/exp.jl:27]:
```suggestion
    x = constant([1, 2, 3])
```

File: src/problem_depot/problems/exp.jl [hunk 3]
```diff
@@ -22,6 +22,19 @@
...
+    p = minimize(sum(exp(x)) + y, y >= 0; numeric_type = T)
+    if test
+        @test problem_vexity(p) == ConvexVexity()
+    end
+    handle_problem!(p)
+    if test
+        @test p.optval ≈ sum(exp.([1,2,3])) atol = atol rtol = rtol
```
Comment (ericphanson) [inline src/problem_depot/problems/exp.jl:34]:
```suggestion
        @test p.optval ≈ sum(exp.([1, 2, 3])) atol = atol rtol = rtol
```

File: src/atoms/ExpAtom.jl [hunk 4]
```diff
@@ -34,9 +34,43 @@ function new_conic_form!(context::Context{T}, e::ExpAtom) where {T}
...
+    # since `ExpAtom` is restricted to `sign(x)` being real, `x_tape` is either `Vector{T}` or `SparseTape{T}`.
+    # The `Vector{T}` case happens when `x` is a constant (or a function of a constant).
+    # In this case, we can just take `exp` directly.
+    if x_tape isa Vector
```
Comment (ericphanson) [inline src/atoms/ExpAtom.jl:51]:
```suggestion
    # since `ExpAtom` is restricted to `sign(x)` being real, `x_tape` is either `SPARSE_VECTOR{T}` or `SparseTape{T}`.
    # The `SPARSE_VECTOR{T}` case happens when `x` is a constant (or a function of a constant).
    # In this case, we can just take `exp` directly.
    if x_tape isa SPARSE_VECTOR
```

**Comment by odow** [comment]
Closing in favor of #618
